### ISSUE TITLE
Unable to mount datasets due to error 17 (cannot mount due to Invalid argument)

### ISSUE THREAD

**Issue by koalaeagle** [zfsonlinux/zfs#6471]
### System information

Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       | Arch
Distribution Version    | Rolling
Linux Kernel                 | 4.12.3-1-vfio #1 SMP PREEMPT Sun Jul 30 16:04:48 UTC 2017 x86_64 GNU/Linux
Architecture                 | x86_64
ZFS Version                  | 0.7.0-rc4
SPL Version                  | 0.7.0-rc4

### Describe the problem you're observing
I am unable to mount my datasets properly. I am receiving error 17 (which I was unable to find information on). The pool appears healthy and the data seems like it is still there, however, I can't figure out how to get it to mount and am not sure how to troubleshoot further.
### Describe how to reproduce the problem
I do not think this would be easy to reproduce and am not sure how to reproduce it on another system since I am not sure what is causing the problem. I can reproduce this on my own system by exporting my pool and importing it again.
Things that may be useful to know:
I did recently run "zpool upgrade".
I did recently scrub the pool after 2 months without scrubbing (did not realize my cron syntax was messed up)
I did have some docker mounts pointing inside of the zfs mount which created files after it failed to mount. I have made sure all of these files were removed, disabled docker, and also moved the mountpoint to a location that has never been used on my system before and is not in use by anything (/mnt/tank).
Datasets with names have been censored.

Basic troubleshooting:
The machine has been rebooted 10+ times.
Moved the mountpoint to a never used before location to avoid conflicting mounts/files.
Attempted zfs mount -a which yields the same error 17 result as below.
Set mountpoint=legacy and mounting through fstab. This did not display the error 17 but the end result was the same (just shadow directories and no data).
I am away from the machine right now but plan on seeing the result when booting into a live cd and importing the pool.
### Include any warning/errors/backtraces from the system logs

```
zpool status
pool: tank
 state: ONLINE
  scan: scrub repaired 0B in 20h18m with 0 errors on Sun Jul 23 17:16:59 2017
config:

        NAME                                          STATE     READ WRITE CKSUM
        tank                                          ONLINE       0     0     0
          raidz2-0                                    ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N6RUJ0C7  ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N4SV2A3F  ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N3HZTVAU  ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N7LVCR0Z  ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N4EZ7X9E  ONLINE       0     0     0
            ata-WDC_WD30EFRX-68EUZN0_WD-WCC4N5ECAC88  ONLINE       0     0     0

errors: No known data errors
```
```
zpool export tank
rm -r /mnt/tank
zpool import tank
filesystem 'tank/ACD' can not be mounted due to error 17
cannot mount 'tank/ACD': Invalid argument
filesystem 'tank/Ansible' can not be mounted due to error 17
cannot mount 'tank/Ansible': Invalid argument
filesystem 'tank/Backup' can not be mounted due to error 17
cannot mount 'tank/Backup': Invalid argument
filesystem 'tank/ClientData' can not be mounted due to error 17
cannot mount 'tank/ClientData': Invalid argument
cannot mount '/mnt/tank/Data': directory is not empty
filesystem 'tank/Data/GiteaData' can not be mounted due to error 17
cannot mount 'tank/Data/GiteaData': Invalid argument
filesystem 'tank/Data/aur' can not be mounted due to error 17
cannot mount 'tank/Data/aur': Invalid argument
filesystem 'tank/Downloads' can not be mounted due to error 17
cannot mount 'tank/Downloads': Invalid argument
filesystem 'tank/Drone' can not be mounted due to error 17
cannot mount 'tank/Drone': Invalid argument
filesystem 'tank/OtherPerson' can not be mounted due to error 17
cannot mount 'tank/OtherPerson': Invalid argument
filesystem 'tank/FilingCabinet' can not be mounted due to error 17
cannot mount 'tank/FilingCabinet': Invalid argument
filesystem 'tank/Games' can not be mounted due to error 17
cannot mount 'tank/Games': Invalid argument
filesystem 'tank/KeePass' can not be mounted due to error 17
cannot mount 'tank/KeePass': Invalid argument
cannot mount '/mnt/tank/Learning': directory is not empty
filesystem 'tank/Learning/Certifications' can not be mounted due to error 17
cannot mount 'tank/Learning/Certifications': Invalid argument
filesystem 'tank/Learning/School' can not be mounted due to error 17
cannot mount 'tank/Learning/School': Invalid argument
filesystem 'tank/Media' can not be mounted due to error 17
cannot mount 'tank/Media': Invalid argument
filesystem 'tank/Movies' can not be mounted due to error 17
cannot mount 'tank/Movies': Invalid argument
filesystem 'tank/Music' can not be mounted due to error 17
cannot mount 'tank/Music': Invalid argument
filesystem 'tank/NextCloudData' can not be mounted due to error 17
cannot mount 'tank/NextCloudData': Invalid argument
filesystem 'tank/Organization' can not be mounted due to error 17
cannot mount 'tank/Organization': Invalid argument
filesystem 'tank/Photos' can not be mounted due to error 17
cannot mount 'tank/Photos': Invalid argument
filesystem 'tank/Podcasts' can not be mounted due to error 17
cannot mount 'tank/Podcasts': Invalid argument
filesystem 'tank/Programs' can not be mounted due to error 17
cannot mount 'tank/Programs': Invalid argument
cannot mount '/mnt/tank/Person': directory is not empty
filesystem 'tank/Person/Gitea' can not be mounted due to error 17
cannot mount 'tank/Person/Gitea': Invalid argument
filesystem 'tank/Person/Github' can not be mounted due to error 17
cannot mount 'tank/Person/Github': Invalid argument
filesystem 'tank/Person/Org' can not be mounted due to error 17
cannot mount 'tank/Person/Org': Invalid argument
filesystem 'tank/Sickmp4' can not be mounted due to error 17
cannot mount 'tank/Sickmp4': Invalid argument
filesystem 'tank/TV Shows' can not be mounted due to error 17
cannot mount 'tank/TV Shows': Invalid argument
filesystem 'tank/VMBackup' can not be mounted due to error 17
cannot mount 'tank/VMBackup': Invalid argument
filesystem 'tank/VMs' can not be mounted due to error 17
cannot mount 'tank/VMs': Invalid argument
filesystem 'tank/VMsExt' can not be mounted due to error 17
cannot mount 'tank/VMsExt': Invalid argument
filesystem 'tank/isos' can not be mounted due to error 17
cannot mount 'tank/isos': Invalid argument
filesystem 'tank/pool' can not be mounted due to error 17
cannot mount 'tank/pool': Invalid argument
```
```
du -sh /mnt/tank/*
512     ACD
512     Ansible
512     Backup
512     ClientData
1.5K    Data
512     Downloads
512     Drone
512     OtherPerson
512     FilingCabinet
512     Games
512     isos
512     KeePass
1.5K    Learning
512     Media
512     Movies
512     Music
512     NextCloudData
512     Organization
512     Photos
512     Podcasts
512     pool
512     Programs
2.0K    Person
512     Sickmp4
512     TV Shows
512     VMBackup
512     VMs
512     VMsExt
```
```
zfs list
NAME                           USED  AVAIL  REFER  MOUNTPOINT
tank                          10.0T   487G   224K  /mnt/tank
tank/ACD                       192K   487G   192K  /mnt/tank/ACD
tank/Ansible                   192K   487G   192K  /mnt/tank/Ansible
tank/Backup                    925G   487G   921G  /mnt/tank/Backup
tank/ClientData                565G   487G   565G  /mnt/tank/ClientData
tank/Data                     9.96G   487G  9.96G  /mnt/tank/Data
tank/Data/GiteaData            192K   487G   192K  /mnt/tank/Data/GiteaData
tank/Data/aur                   272K   487G   272K  /mnt/tank/Data/aur
tank/Downloads                86.3G   487G  86.3G  /mnt/tank/Downloads
tank/Drone                     352K   487G   352K  /mnt/tank/Drone
tank/OtherPerson             15.3G   487G  15.3G  /mnt/tank/OtherPerson
tank/FilingCabinet            1.06G   487G  1.06G  /mnt/tank/FilingCabinet
tank/Games                    1.56G   487G  1.56G  /mnt/tank/Games
tank/KeePass                  5.28M   487G   280K  /mnt/tank/KeePass
tank/Learning                  623K   487G   192K  /mnt/tank/Learning
tank/Learning/Certifications   240K   487G   240K  /mnt/tank/Learning/Certifications
tank/Learning/School           192K   487G   192K  /mnt/tank/Learning/School
tank/Media                     192K   487G   192K  /mnt/tank/Media
tank/Movies                    344K   487G   344K  /mnt/tank/Movies
tank/Music                    16.2G   487G  16.2G  /mnt/tank/Music
tank/NextCloudData            3.21T   487G  1.81T  /mnt/tank/NextCloudData
tank/Organization             1.58M   487G   328K  /mnt/tank/Organization
tank/Photos                    495M   487G   495M  /mnt/tank/Photos
tank/Podcasts                  192K   487G   192K  /mnt/tank/Podcasts
tank/Programs                  192K   487G   192K  /mnt/tank/Programs
tank/Person                    1.64T   487G  1.64T  /mnt/tank/Person
tank/Person/Gitea               192K   487G   192K  /mnt/tank/Person/Gitea
tank/Person/Github              192K   487G   192K  /mnt/tank/Person/Github
tank/Person/Org                 192K   487G   192K  /mnt/tank/Person/Org
tank/Sickmp4                  91.7G   487G  91.7G  /mnt/tank/Sickmp4
tank/TV Shows                  426G   487G   426G  /mnt/tank/TV Shows
tank/VMBackup                 24.4G   487G  24.4G  /mnt/tank/VMBackup
tank/VMs                       204M   487G   204M  /mnt/tank/VMs
tank/VMsExt                    180G   487G   180G  /mnt/tank/VMsExt
tank/isos                      192K   487G   192K  /mnt/tank/isos
tank/pool                     2.87T   487G  2.87T  /mnt/tank/pool
```
```
[ 5626.775499] ------------[ cut here ]------------
[ 5626.775504] WARNING: CPU: 6 PID: 6337 at lib/kobject.c:240 kobject_add_internal+0x265/0x2b0
[ 5626.775505] Modules linked in: nls_utf8 ntfs ebtable_filter ebtables ip6table_filter ip6_tables fuse bridge stp llc xt_pkttype tun xt_tcpudp iptable_filter zfs(PO) zunicode(PO) zavl(PO) icp(PO) intel_rapl x86_pkg_temp_thermal intel_powerclamp hid_logitech_hidpp coretemp joydev mousedev kvm_intel mei_wdt iTCO_wdt input_leds iTCO_vendor_support i915 kvm evdev mac_hid crct10dif_pclmul crc32_pclmul crc32c_intel ghash_clmulni_intel drm_kms_helper hid_logitech_dj uas pcbc drm nls_iso8859_1 aesni_intel nls_cp437 aes_x86_64 crypto_simd glue_helper cryptd e1000e intel_gtt zcommon(PO) intel_cstate vfat syscopyarea znvpair(PO) fat sysfillrect intel_rapl_perf sysimgblt spl(O) mei_me ptp pcspkr i2c_i801 fb_sys_fops i2c_algo_bit mei shpchp lpc_ich pps_core ie31200_edac snd_soc_rt5640 snd_soc_rl6231 snd_soc_ssm4567
[ 5626.775570]  battery snd_soc_core nuvoton_cir snd_compress rc_core snd_pcm_dmaengine snd_pcm elan_i2c snd_soc_sst_acpi snd_timer snd_soc_sst_match snd soundcore i2c_hid ac97_bus video i2c_designware_platform 8250_dw i2c_designware_core tpm_tis tpm_tis_core spi_pxa2xx_platform tpm acpi_pad button hid_generic nfsd sch_fq_codel auth_rpcgss oid_registry nfs_acl lockd grace sunrpc ip_tables x_tables ext4 crc16 jbd2 fscrypto mbcache usb_storage usbhid hid sd_mod ahci xhci_pci libahci ehci_pci xhci_hcd ehci_hcd libata usbcore usb_common scsi_mod serio sdhci_acpi sdhci led_class mmc_core loop vfio_pci irqbypass vfio_virqfd vfio_iommu_type1 vfio
[ 5626.775629] CPU: 6 PID: 6337 Comm: mount.zfs Tainted: P        W  O    4.12.3-1-vfio #1
[ 5626.775630] Hardware name: To Be Filled By O.E.M. To Be Filled By O.E.M./C226M WS, BIOS P2.20 08/17/2015
[ 5626.775632] task: ffff93673a92c740 task.stack: ffffa7ec47e84000
[ 5626.775635] RIP: 0010:kobject_add_internal+0x265/0x2b0
[ 5626.775637] RSP: 0018:ffffa7ec47e87b10 EFLAGS: 00010282
[ 5626.775639] RAX: 0000000000000078 RBX: ffff936a490f5010 RCX: 0000000000000006
[ 5626.775641] RDX: 0000000000000000 RSI: 0000000000000082 RDI: ffff936d6eb8dcc0
[ 5626.775643] RBP: ffffa7ec47e87b40 R08: 0000000000000001 R09: 0000000000008ae8
[ 5626.775644] R10: ffffec9bc8f07800 R11: 0000000000008ae8 R12: 0000000000000000
[ 5626.775646] R13: ffff936d43e19c00 R14: 00000000ffffffef R15: 0000000000000000
[ 5626.775648] FS:  00007fa9b80cc7c0(0000) GS:ffff936d6eb80000(0000) knlGS:0000000000000000
[ 5626.775650] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 5626.775651] CR2: 00007fa9b6216ef0 CR3: 00000004d684a000 CR4: 00000000001426e0
[ 5626.775653] Call Trace:
[ 5626.775657]  kobject_add+0x71/0xd0
[ 5626.775662]  ? refcount_inc+0x9/0x30
[ 5626.775667]  device_add+0x151/0x640
[ 5626.775672]  device_create_groups_vargs+0xe0/0xf0
[ 5626.775676]  device_create_vargs+0x16/0x20
[ 5626.775680]  bdi_register_va.part.11+0x28/0x170
[ 5626.775683]  bdi_register_va+0x1b/0x20
[ 5626.775685]  super_setup_bdi_name+0x87/0xe0
[ 5626.775720]  ? dmu_objset_rele+0x42/0x60 [zfs]
[ 5626.775760]  ? dsl_prop_get+0x7b/0xb0 [zfs]
[ 5626.775767]  ? spl_kmem_free+0x33/0x40 [spl]
[ 5626.775805]  zfs_domount+0x110/0x370 [zfs]
[ 5626.775843]  ? zpl_mount+0x50/0x50 [zfs]
[ 5626.775881]  ? zpl_mount+0x50/0x50 [zfs]
[ 5626.775919]  zpl_fill_super+0x2c/0x40 [zfs]
[ 5626.775923]  mount_nodev+0x4d/0xa0
[ 5626.775961]  zpl_mount+0x34/0x50 [zfs]
[ 5626.775964]  mount_fs+0x38/0x140
[ 5626.775970]  vfs_kern_mount+0x67/0x110
[ 5626.775974]  do_mount+0x1dd/0xc60
[ 5626.775977]  ? __check_object_size+0xb3/0x190
[ 5626.775980]  ? refcount_dec_not_one+0x31/0x50
[ 5626.775985]  SyS_mount+0x83/0xd0
[ 5626.775989]  entry_SYSCALL_64_fastpath+0x1a/0xa5
[ 5626.775991] RIP: 0033:0x7fa9b73915fa
[ 5626.775993] RSP: 002b:00007ffdcdd098d8 EFLAGS: 00000206 ORIG_RAX: 00000000000000a5
[ 5626.775996] RAX: ffffffffffffffda RBX: 00000000024ac060 RCX: 00007fa9b73915fa
[ 5626.775997] RDX: 00000000004023e6 RSI: 00007ffdcdd0da00 RDI: 00007ffdcdd107b5
[ 5626.775999] RBP: 00000000024ad3b0 R08: 00007ffdcdd0aa00 R09: 0000000000000003
[ 5626.776001] R10: 0000000000200000 R11: 0000000000000206 R12: 00007ffdcdd0aa00
[ 5626.776002] R13: 0000000000000005 R14: 0000000000000000 R15: 0000000000000000
[ 5626.776006] Code: 48 8b 7b 20 49 89 c5 48 85 ff 0f 84 44 fe ff ff e9 01 fe ff ff 48 8b 13 48 c7 c6 f0 a5 84 8c 48 c7 c7 d0 14 96 8c e8 15 32 e5 ff <0f> ff e9 96 fe ff ff 0f 0b 0f 0b 0f 0b 0f ff eb 9f 0f ff eb 92 
[ 5626.776061] ---[ end trace baeb0e238956d422 ]---
```
The mount on boot of course failed with the same result:
```
systemctl status zfs-mount
zfs-mount.service - Mount ZFS filesystems
   Loaded: loaded (/usr/lib/systemd/system/zfs-mount.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Sun 2017-08-06 18:59:57 EDT; 1h 50min ago
  Process: 1276 ExecStart=/usr/bin/zfs mount -a (code=exited, status=1/FAILURE)
 Main PID: 1276 (code=exited, status=1/FAILURE)

Aug 06 18:59:57 archone zfs[1276]: filesystem 'tank/VMsExt' can not be mounted due to error 17
Aug 06 18:59:57 archone zfs[1276]: cannot mount 'tank/VMsExt': Invalid argument
Aug 06 18:59:57 archone zfs[1276]: filesystem 'tank/isos' can not be mounted due to error 17
Aug 06 18:59:57 archone zfs[1276]: cannot mount 'tank/isos': Invalid argument
Aug 06 18:59:57 archone zfs[1276]: filesystem 'tank/pool' can not be mounted due to error 17
Aug 06 18:59:57 archone zfs[1276]: cannot mount 'tank/pool': Invalid argument
Aug 06 18:59:57 archone systemd[1]: zfs-mount.service: Main process exited, code=exited, status=1/FAILURE
Aug 06 18:59:57 archone systemd[1]: Failed to start Mount ZFS filesystems.
Aug 06 18:59:57 archone systemd[1]: zfs-mount.service: Unit entered failed state.
Aug 06 18:59:57 archone systemd[1]: zfs-mount.service: Failed with result 'exit-code'.
```

**Comment by kpande** [comment]
reverting #6398 fixes this issue.

**Comment by loli10K** [comment]
> reverting #6398 fixes this issue.

According to the issue description this bug occurs with zfs-0.7.0-rc4 which doesn't even include that commit.

**Comment by dasJ** [comment]
Had the same problem and fixed it temporarily by downgrading to linux 4.9.6-1

**Comment by dasJ** [comment]
Oh, my root (one SSD with TRIM patch) mounted well, but my home (RAID Z1) didn't mount with the error above

**Comment by loli10K** [comment]
Fixed in https://github.com/zfsonlinux/zfs/commit/3e6c9433474f0b6fc4880549c851bfb151c8c719 (not included in zfs-0.7.0-rc4), a Linux downgrade or ZFS upgrade will likely fix this, closing.
